snuba/api.py

```diff
@@ -162,16 +162,80 @@ def query(validated_body=None, timer=None):
     )
 
 
+def format_query(dataset, body, table, prewhere_conditions, final) -> str:
+    """Generate a SQL string from the parameters."""
+    aggregate_exprs = [util.column_expr(dataset, col, body, alias, agg) for (agg, col, alias) in body['aggregations']]
+    groupby = util.to_list(body['groupby'])
+    group_exprs = [util.column_expr(dataset, gb, body) for gb in groupby]
+    selected_cols = [util.column_expr(dataset, util.tuplify(colname), body) for colname in body.get('selected_columns', [])]
+    select_clause = u'SELECT {}'.format(', '.join(group_exprs + aggregate_exprs + selected_cols))
+
+    from_clause = u'FROM {}'.format(table)
+    if final:
+        from_clause = u'{} FINAL'.format(from_clause)
+    if 'sample' in body:
+        from_clause = u'{} SAMPLE {}'.format(from_clause, body['sample'])
+
+    join_clause = ''
+    if 'arrayjoin' in body:
+        join_clause = u'ARRAY JOIN {}'.format(body['arrayjoin'])
+
+    where_clause = ''
+    if body['conditions']:
+        where_clause = u'WHERE {}'.format(util.conditions_expr(dataset, body['conditions'], body))
+
+    prewhere_clause = ''
+    if prewhere_conditions:
+        prewhere_clause = u'PREWHERE {}'.format(util.conditions_expr(dataset, prewhere_conditions, body))
+
+    group_clause = ''
+    if groupby:
+        group_clause = 'GROUP BY ({})'.format(', '.join(util.column_expr(dataset, gb, body) for gb in groupby))
+        if body.get('totals', False):
+            group_clause = '{} WITH TOTALS'.format(group_clause)
+
+    having_clause = ''
+    having_conditions = body.get('having', [])
+    if having_conditions:
+        assert groupby, 'found HAVING clause with no GROUP BY'
+        having_clause = u'HAVING {}'.format(util.conditions_expr(dataset, having_conditions, body))
+
+    order_clause = ''
+    if body.get('orderby'):
+        orderby = [util.column_expr(dataset, util.tuplify(ob), body) for ob in util.to_list(body['orderby'])]
+        orderby = [u'{} {}'.format(ob.lstrip('-'), 'DESC' if ob.startswith('-') else 'ASC') for ob in orderby]
+        order_clause = u'ORDER BY {}'.format(', '.join(orderby))
+
+    limitby_clause = ''
+    if 'limitby' in body:
+        limitby_clause = 'LIMIT {} BY {}'.format(*body['limitby'])
```

Comment:
This was moved, but I'm pretty sure this won't actually work because it's not using `column_expr`. I didn't test it.

Reply:
moved from where ?

Reply:
https://github.com/getsentry/snuba/blob/2d55ce81766fbc957822bfcea7b47b2156c4b93d/snuba/api.py#L370-L372

Almost everything here should be a move out of `parse_and_run_query`. Sorry if this wasn't clear from the description.